### PR TITLE
split up tests

### PR DESCRIPTION
**Pull request by ExpandingMan**
This is a re-do of #1824 for post 0.13.

To reiterate, the goal here is to roughly group the tests by the functionality they are testing and separate them into a large number of files which should be possible to run entirely independently.  Again the way I have grouped these tests may not always make a lot of sense, it seems likely they'll get shuffled around more in the future (or before this gets merged) but this seems like a decent start.

There is still a major issue that I've so far been unable to resolve: a setfault in `arrays.jl`, which I've seen before when trying to split up the tests.  It sure looks to me like the fact that it segfaults is an enzyme bug, as I don't see any reason it should do that, especially merely as a result of being in a different file.

As of writing, everything else in here passes as expected for me (x86 linux).

### CODE REVIEW

File: test/arrays.jl [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+    xmean = mean(x_, dims=(1, 2, 5))
+    xvar = var(x_, corrected=false, mean=xmean, dims=(1, 2, 5))
+
+    return sum(abs2, bc2_affine_normalize(identity, x_, xmean, xvar, scale_, bias_, 1e-5))
+end
+
+#FUCK: I'm *still* getting segfaults here????
+#=
+@testset "Broadcast noalias" begin
```
Comment (ExpandingMan) [inline test/arrays.jl:246]: The segfault I can't explain happens in here.

File: test/higherorder.jl [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    x = [1.0]
+    v = [-1.0]
+    hv = make_zero(v)
+    tmp = similar(x)
+
+    #WARN: this fails with an assertion error from somewhere deep inside enzyme/compiler and I 
+    # have absolutely no idea why
+    # definitely fails when file is run in isolation, not sure about otherwise
+    f_hvp!(hv, x, v, tmp)
```
Comment (ExpandingMan) [inline test/higherorder.jl:71]: The assertion error happens in this call.

Comment (ExpandingMan) [reply test/higherorder.jl:69]: Turns out this was due to `LinearAlgebra` not being imported, and `dot` not being defined.  The error message was unrecognizable to me, but there was clearly an error introduced by me here, now fixed.

**Comment by wsmoses** [comment]
open a mwe issue for each of these?

**Comment by ExpandingMan** [comment]
#1869 and #1870

(the former is clearly a real bug, the latter is just confusing error messages and probably not a bug)

**Comment by wsmoses** [comment]
I think we need to fix the segfault before this goes in.

Unfortunately I can't reproduce locally. Can you setup a tmate instance with it on it or something, and send me via slack

**Comment by ExpandingMan** [comment]
Ok.  I don't think I've ever let the segfaulting test run in CI/CD, let's see what that does first, I've just uncommented it.

**Comment by ExpandingMan** [comment]
Frankly that could take forever, there is quite a lot of stuff to get split up.  I'd like to get it in, but if you think it's going to be that difficult, I'll leave it to you (if this is indeed something you feel is worth doing).  Even if I did this piecemeal I wouldn't know when to do what.

**Comment by wsmoses** [comment]
maybe do like 3-4 in a first pass (and move out utils)?

**Comment by ExpandingMan** [comment]
Could we maybe just try to keep track of which tests are added between the PR going up and it being merged?  Would it really be that hard?

**Comment by wsmoses** [comment]
I mean sure if you want, but also bear in mind that moving everything together means it'll take longer to review.....which means another test PR might get in before then....necessitating another review...

**Pull request closed without merging**